pkg/bff/organization.go

```diff
@@ -13,6 +15,145 @@ import (
 	storeerrors "github.com/trisacrypto/directory/pkg/store/errors"
 )
 
+// CreateOrganization creates a new organization in the database. This endpoint returns
+// an error if the organization already exists and the user is assigned to it. The user
+// must have the create:organizations permission to perform this action.
+func (s *Server) CreateOrganization(c *gin.Context) {
+	var (
+		err  error
+		user *management.User
+	)
+
+	// Fetch the user from the context
+	if user, err = auth.GetUserInfo(c); err != nil {
+		log.Error().Err(err).Msg("create organization handler requires user info; expected middleware to return 401")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not identify user to create organization"))
+		return
+	}
+
+	// Load the user app metadata to check their organization assignments
+	appdata := &auth.AppMetadata{}
+	if err = appdata.Load(user.AppMetadata); err != nil {
+		log.Error().Err(err).Msg("could not parse user app metadata")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not parse user app metadata"))
+		return
+	}
+
+	// Unmarshal the params from the POST request
+	params := &api.OrganizationParams{}
+	if err := c.ShouldBind(params); err != nil {
+		log.Warn().Err(err).Msg("could not bind request")
+		c.JSON(http.StatusBadRequest, &api.Reply{Error: err.Error()})
+		return
+	}
+
+	// Name is a required paramater
+	if params.Name == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("must provide name in request params"))
+		return
+	}
+
+	// Domain is a required paramater
+	domain := normalizeDomain(params.Domain)
+	if domain == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("must provide domain in request params"))
+		return
+	}
+
+	// Don't allow the user to create duplicate organizations
+	// TODO: Should we do a universal check against the database using an index?
+	for _, id := range appdata.GetOrganizations() {
+		if org, err := s.OrganizationFromID(id); err != nil {
+			log.Error().Err(err).Str("org_id", id).Msg("could not retrieve organization from database")
+		} else if org.Domain == domain {
+			c.JSON(http.StatusConflict, api.ErrorResponse("organization with domain name already exists"))
+			return
+		}
+	}
```

Comment:
It might be helpful to put the search as a `Contains` method on `appdata` -- it seems like something we will want to do frequently and it will make it easier to optimize in the future (e.g. binary search with an ordered list of organizations).

Reply:
Agreed!